cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberService.java

```diff
@@ -402,6 +402,21 @@ private void handleProgramEvent(ProgramRunId programRunId, ProgramRunStatus prog
         recordedRunRecord = appMetadataStore.recordProgramResumed(programRunId, messageIdBytes, resumeTime);
         writeToHeartBeatTable(recordedRunRecord, resumeTime, programHeartbeatTable);
         break;
+      case STOPPING:
+        long stoppingTimestamp = getTimeSeconds(notification.getProperties(), ProgramOptionConstants.STOPPING_TIME);
+        if (stoppingTimestamp == -1) {
+          LOG.warn("Ignore program stopping notification for program {} without {} specified, {}",
+                   programRunId, ProgramOptionConstants.STOPPING_TIME, notification);
+          return;
+        }
+        long gracefulShutdownPeriod = getTimeSeconds(notification.getProperties(),
+                                                     ProgramOptionConstants.GRACEFUL_SHUTDOWN_PERIOD_IN_SECS);
+        long stoppingTimeoutTs = (Long.MAX_VALUE - gracefulShutdownPeriod >= stoppingTimestamp) ? Long.MAX_VALUE :
```

Comment:
Is this supposed to be the timestamp in millis for the stop timeout?

Reply:
This refers to the end time stamp at which the program should be stopped. The timeunit is seconds. The suffix `Ts` refers to it being a timestamp and all the timestamps in `RunRecordDetail` are in seconds including `stoppingTimestamp`

Reply:
Changed graceful period to int as discussed offline

Reply:
Also changed `stoppingTimeoutTs` to `terminateTimestamp` similar to the way other timestamp variables are named. Since all timestamp long variables have a timeunit of seconds, I did not explicitly add the unit in the variable name